x-pack/plugins/security/server/anonymous_access/anonymous_access_service.ts

```diff
@@ -0,0 +1,140 @@
+import { addSpaceIdToPath } from '../../../spaces/common';
+import type { SpacesServiceStart } from '../../../spaces/server';
+import { AUTH_PROVIDER_HINT_QUERY_STRING_PARAMETER } from '../../common/constants';
+import { AnonymousAuthenticationProvider } from '../authentication';
+import type { AuthorizationServiceSetup } from '../authorization';
+import type { ConfigType } from '../config';
+
+export interface AnonymousAccessServiceSetup {
+  readonly isAnonymousAccessEnabled: boolean;
+}
+
+export interface AnonymousAccessServiceStart {
+  readonly isAnonymousAccessEnabled: boolean;
+  // We cannot use `ReadonlyMap` just yet: https://github.com/microsoft/TypeScript/issues/16840
+  readonly accessURLParameters: Readonly<Map<string, string>> | null;
+  isSavedObjectTypeAccessibleAnonymously: (
+    request: KibanaRequest,
+    savedObjectType: string
+  ) => Promise<boolean>;
+}
+
+interface AnonymousAccessServiceStartParams {
+  authz: Pick<AuthorizationServiceSetup, 'actions' | 'checkSavedObjectsPrivilegesWithRequest'>;
+  basePath: IBasePath;
+  spaces?: SpacesServiceStart;
+}
+
+/**
+ * Service that manages various aspects of the anonymous access.
+ */
+export class AnonymousAccessService {
+  constructor(private readonly logger: Logger, private readonly getConfig: () => ConfigType) {}
+
+  setup(): AnonymousAccessServiceSetup {
+    return {
+      isAnonymousAccessEnabled: this.isAnonymousAccessEnabled(),
+    };
+  }
+
+  start({
+    authz,
+    basePath,
+    spaces,
+  }: AnonymousAccessServiceStartParams): AnonymousAccessServiceStart {
+    const config = this.getConfig();
+    const anonymousProvider = config.authc.sortedProviders.find(
+      ({ type }) => type === AnonymousAuthenticationProvider.type
+    );
+    // We don't need to add any special parameters to the URL if any of the following is true:
+    // * anonymous provider isn't enabled
+    // * anonymous provider is enabled, but it's a default authentication mechanism
+    const anonymousIsDefault =
+      !config.authc.selector.enabled && anonymousProvider === config.authc.sortedProviders[0];
+
+    return {
+      isAnonymousAccessEnabled: this.isAnonymousAccessEnabled(),
+      accessURLParameters:
+        anonymousProvider && !anonymousIsDefault
+          ? Object.freeze(
+              new Map([[AUTH_PROVIDER_HINT_QUERY_STRING_PARAMETER, anonymousProvider.name]])
+            )
+          : null,
+      isSavedObjectTypeAccessibleAnonymously: async (request, savedObjectType) => {
+        this.logger.debug(`Checking if Saved Object is accessible anonymously.`);
+
+        // We should use credentials of the anonymous service account instead of credentials of the
+        // current user to figure out if the specified Saved Object type can be accessed anonymously.
+        const anonymousAuthorizationHeader = this.createAnonymousAuthorizationHeader();
+        const fakeAnonymousRequest = KibanaRequest.from(({
```

Comment:
__question:__ do we have any better option to do what I do right now?

Reply:
Not really 🙁. Core support for something better is being discussed in https://github.com/elastic/kibana/issues/39430, but for the time being this is the generally accepted short-term solution.

Reply:
Got it, thanks for confirming!